libcst/codemod/commands/tests/test_rename.py

```diff
@@ -0,0 +1,202 @@
+
+        before = """
+            from foo import bar as bla
+
+            def test() -> None:
+                bla(5)
+        """
+        after = """
+            from baz import qux as bla
+
+            def test() -> None:
+                bla(5)
+        """
+
+        self.assertCodemod(
+            before,
+            after,
+            orig_module="foo",
+            orig_object="bar",
+            new_module="baz",
+            new_object="qux",
+        )
+
+    def test_rename_attr(self) -> None:
+
+        before = """
+            import a.b
+
+            def test() -> None:
+                a.b.c(5)
+        """
+        after = """
+            import d.e
+
+            def test() -> None:
+                d.e.f(5)
+        """
+
+        self.assertCodemod(
+            before,
+            after,
+            orig_module="a.b",
+            orig_object="c",
+            new_module="d.e",
+            new_object="f",
+        )
+
+    def test_rename_attr_asname(self) -> None:
+
+        before = """
+            import foo as bar
+
+            def test() -> None:
+                bar.qux(5)
+        """
+        after = """
+            import baz as bar
+
+            def test() -> None:
+                bar.quux(5)
+        """
+
+        self.assertCodemod(
+            before,
+            after,
+            orig_module="foo",
+            orig_object="qux",
+            new_module="baz",
+            new_object="quux",
+        )
+
+    def test_rename_module(self) -> None:
+        before = """
+            from a import b
+
+            class Foo(b):
+                pass
+        """
+        after = """
+            from c import b
+
+            class Foo(b):
+                pass
+        """
+
+        self.assertCodemod(
+            before,
+            after,
+            orig_module="a",
+            orig_object="b",
+            new_module="c",
+            new_object="b",
+        )
+
+    def test_rename_local_variable(self) -> None:
+        before = """
+            x = 5
+            y = 5 + x
+        """
+        after = """
+            z = 5
+            y = 5 + z
+        """
+
+        self.assertCodemod(
+            before, after, orig_object="x", new_object="z",
+        )
+
+    def test_rename_local_object_attr(self) -> None:
+        before = """
+            x = Foo()
+            x.bar = 5
+        """
+        after = """
+            x = Foo()
+            x.baz = 5
+        """
+        self.assertCodemod(
+            before, after, orig_object="x.bar", new_object="x.baz",
+        )
+
+    def test_module_does_not_change(self) -> None:
+        before = """
+            from a import b
+
+            class Foo(b):
+                pass
+        """
+        after = """
+            from a import c
+
+            class Foo(c):
+                pass
+        """
+        self.assertCodemod(
+            before, after, orig_module="a", orig_object="b", new_object="c"
```

Comment:
Reading this, I was initially a bit confused that we implicitly assume `new_module` is "a". It could be more obvious if we required the `new_module` parameter any time there is an `old_module` parameter.

Reply:
We'll always need a `orig_module` to locate the import module of `orig_object`. When `new_module` is not provided, it means the developer don't want to rename the module name. So at least one of `new_module` or `new_object` should be provided in a rename use case.

Reply:
I have allowed the omission of both for the case of renaming a local variable. Do we want to disallow this altogether?

Reply:
Renaming a local variable is a very cool feature! 

The reservation I have with the current implementation is that you oftentimes a local variable name is repeated within multiple classes, methods, or functions within a module. I don't think it makes sense to rename all of the local variables in the same manner.

So, for renaming a local variable, I imagine that you want to target a scope that may be more granular than a file / module. 

This is why I was thinking that it could make sense to have a different interface for renaming local variables, and hence a separate codemod command.

Reply:
There are multiple levels of "local"; here it just means module-local (not imported). This could mean a module-level "global" variable, which would be reasonable to want to rename module-wide.

Reply:
I looked a bit more into how `QualifiedNameProvider` works. Presumably the reason why this is the case is because the qualified name of a function local `y` in `x`

```
def x():
  y = 1
  ...
```
is `x.<locals>.y` and not simply `y`. Does this mean that we can rename function or class locals by correctly providing the right qualified name?

Reply:
Also, we generally design codemod commands to run over an entire codebase and not a single module or file. Renaming local variables (whether module-local or not) only really makes sense to run on a single module - it's unlikely you want to rename a local variable across the codebase.

The question I have is: should we provide a separate interface for renames of local variables that takes the target module as well?